### ISSUE TITLE
Recreation loss for goal autoencoder

### ISSUE THREAD

**Issue by dirkmcpherson** [danijar/director#7]
Hi, 

The goal autoencoder's recreation loss in your code is the negative log probability of the world model's representation under the goal decoder's distribution:

`rec = -dec.log_prob(tf.stop_gradient(goal))`

But the paper lists it as the MSE of the decoded state and the original state:

`np.sqrt(dec(feat.detach()) - feat)`

Is the former a better measure of recreation loss than the latter?

Also, you only use the deterministic part of the RSSM as the representation, for training the goal autoencoder (in hierarchy::train_vae_replay). Why only use the deterministic part and not include the stochastic part? 

Apologies if you've written about this somewhere, and thank you for making your extremely interesting work public.

**Comment by dirkmcpherson** [comment]
I think the answer to the first part of my question is that the decoder is using an symlog_mse distribution, so you can use the likelihood of the state under that distribution as a proxy for recreation loss. 

The second part of my question stands though, the RSSM uses [stoch, deter] to find one-hot representations, why doesn't the goal autoencoder?

**Comment by danijar** [comment]
Hi, MSE and gaussian negative log prob are the same up to constant offset and scaling (just like L1 and Laplace etc).

When training the worker in imagination, z is sampled from h so the best the worker can do is reach the correct h. Additionally rewarding it for also reaching the right z just adds noise to the reward because the worker cannot control it.
